otelconf/v0.3.0/log_test.go

```diff
@@ -694,3 +704,217 @@ func TestLogProcessor(t *testing.T) {
+				require.Equal(t, tt.wantErr, err.Error())
+			} else {
+				require.NoError(t, err)
+			}
+
+			logFactory := sdklogtest.RecordFactory{
+				Body: log.StringValue("test"),
+			}
+
+			require.NoError(t, exporter.Export(context.Background(), []sdklog.Record{
+				logFactory.NewRecord(),
+			}))
+
+			// Ensure everything is flushed.
+			require.NoError(t, exporter.Shutdown(context.Background()))
+
+			// verify the reception of the sent data item
+			require.GreaterOrEqual(t, len(col.storage.data), 1)
+		})
+	}
+}
+
+// logsStorage stores uploaded OTLP log data in their proto form.
+type logsStorage struct {
+	dataMu sync.Mutex
+	data   []*lpb.ResourceLogs
+}
+
+// Add adds the request to the Storage.
+func (s *logsStorage) Add(request *collogpb.ExportLogsServiceRequest) {
+	s.dataMu.Lock()
+	defer s.dataMu.Unlock()
+	s.data = append(s.data, request.ResourceLogs...)
+}
+
+// grpcLogsCollector is an OTLP gRPC server that collects all requests it receives.
+type grpcLogsCollector struct {
+	collogpb.UnimplementedLogsServiceServer
+
+	storage *logsStorage
+
+	listener net.Listener
+	srv      *grpc.Server
+}
+
+var _ collogpb.LogsServiceServer = (*grpcLogsCollector)(nil)
+
+// newGRPCLogsCollector returns a *grpcLogsCollector that is listening at the provided
+// endpoint.
+//
+// If endpoint is an empty string, the returned collector will be listening on
+// the localhost interface at an OS chosen port.
+func newGRPCLogsCollector(listener net.Listener, tlsMode string) (*grpcLogsCollector, error) {
+	c := &grpcLogsCollector{
+		storage:  &logsStorage{},
+		listener: listener,
+	}
+
+	srv, err := createGRPCServer(tlsMode)
+	if err != nil {
+		return nil, err
+	}
+
+	c.srv = srv
+	collogpb.RegisterLogsServiceServer(c.srv, c)
+	go func() { _ = c.srv.Serve(c.listener) }()
+
+	return c, nil
+}
+
+// Export handles the export req.
+func (c *grpcLogsCollector) Export(
+	_ context.Context,
+	req *collogpb.ExportLogsServiceRequest,
+) (*collogpb.ExportLogsServiceResponse, error) {
+	c.storage.Add(req)
+
+	return &collogpb.ExportLogsServiceResponse{}, nil
+}
+
+func createGRPCServer(tlsMode string) (*grpc.Server, error) {
+	opts := []grpc.ServerOption{}
+
+	switch tlsMode {
+	case "TLS":
+		tlsCreds, err := credentials.NewServerTLSFromFile("testdata/server-certs/server.crt", "testdata/server-certs/server.key")
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, grpc.Creds(tlsCreds))
+	case "mTLS":
+		cert, err := tls.LoadX509KeyPair("testdata/server-certs/server.crt", "testdata/server-certs/server.key")
+		if err != nil {
+			return nil, err
+		}
+		caCert, err := os.ReadFile("testdata/ca.crt")
+		if err != nil {
+			return nil, err
+		}
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tlsCreds := credentials.NewTLS(&tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientCAs:    caCertPool,
+			ClientAuth:   tls.RequireAndVerifyClientCert,
+		})
+		opts = append(opts, grpc.Creds(tlsCreds))
+	default:
```

Comment:
How about adding an `func() []grpc.ServerOption` field to test cases struct so that we can get rid of conditional logic in tests and add this logic to the test cases?

Reply:
i feel like that will add quite some verbosity to the test cases, but I will give it a try